pkg/controller/database/rdsinstance.go

```diff
@@ -204,18 +205,16 @@ func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.Ext
 	modify := rds.GenerateModifyDBInstanceInput(meta.GetExternalName(cr), patch)
 	var conn managed.ConnectionDetails
 	if cr.Spec.ForProvider.MasterPasswordSecretRef != nil {
-		s := &corev1.Secret{}
-		nn := types.NamespacedName{
-			Name:      cr.Spec.ForProvider.MasterPasswordSecretRef.Name,
-			Namespace: cr.Spec.ForProvider.MasterPasswordSecretRef.Namespace,
-		}
-		if err := e.kube.Get(ctx, nn, s); err != nil {
-			return managed.ExternalUpdate{}, errors.Wrap(err, errGetPasswordSecretFailed)
+		pwd, changed, err := rds.GetPassword(ctx, &e.kube, cr.Spec.ForProvider.MasterPasswordSecretRef, cr.Spec.WriteConnectionSecretToReference)
+		if err != nil {
+			return managed.ExternalUpdate{}, err
 		}
-		conn = managed.ConnectionDetails{
-			runtimev1alpha1.ResourceCredentialsSecretPasswordKey: s.Data[cr.Spec.ForProvider.MasterPasswordSecretRef.Key],
+		if changed {
+			conn = managed.ConnectionDetails{
+				runtimev1alpha1.ResourceCredentialsSecretPasswordKey: []byte(pwd),
+			}
+			modify.MasterUserPassword = aws.String(pwd)
 		}
-		modify.MasterUserPassword = aws.String(string(s.Data[cr.Spec.ForProvider.MasterPasswordSecretRef.Key]))
 	}
 	if _, err = e.client.ModifyDBInstanceRequest(modify).Send(ctx); err != nil {
 		return managed.ExternalUpdate{}, errors.Wrap(err, errModifyFailed)
```

Comment:
Would you be use `GetPassword` in `Create`, too? We wouldn't really check whether password changed but we'd need the desired password and generate a new one only if the returned password is `""`.

Reply:
Wonderful comments, I'll make the according changes and look at the tests you referenced to see what I can do with the clients!